modin/core/dataframe/base/dataframe/utils.py

```diff
@@ -48,3 +53,71 @@ class JoinType(Enum):  # noqa: PR01
     LEFT = "left"
     RIGHT = "right"
     OUTER = "outer"
+
+
+def join_columns(
+    left: pandas.Index,
+    right: pandas.Index,
+    left_on: Union[Iterable[IndexLabel], IndexLabel],
+    right_on: Union[Iterable[IndexLabel], IndexLabel],
+    suffixes: Tuple[str, str],
+) -> Tuple[pandas.Index, Dict[IndexLabel, IndexLabel], Dict[IndexLabel, IndexLabel]]:
+    """
+    Compute resulting columns for the two dataframes being merged.
+
+    Parameters
+    ----------
+    left : pandas.Index
+        Columns of the left frame to join.
+    right : pandas.Index
+        Columns of the right frame to join.
+    left_on : list-like or scalar
+        Column names on which the frames are joined in the left DataFrame.
+    right_on : list-like or scalar
+        Column names on which the frames are joined in the right DataFrame.
+    suffixes : tuple[str, str]
+        A 2-length sequence containing suffixes to append to the intersected columns.
+
+    Returns
+    -------
+    pandas.Index, dict[IndexLabel -> IndexLabel], dict[IndexLabel -> IndexLabel]
+        Returns columns for the resulting frame and mappings of old to new column
+        names for `left` and `right` accordingly.
+    """
+    left_on, right_on = map(
+        lambda x: cast(Iterable[IndexLabel], [x] if is_scalar(x) else x),
+        (left_on, right_on),
+    )
+
+    left_conflicts = set(left) & (set(right) - set(right_on))
+    right_conflicts = set(right) & (set(left) - set(left_on))
+    conflicting_cols = left_conflicts | right_conflicts
+
+    def _get_new_name(col: IndexLabel, suffix: str) -> IndexLabel:
+        if col in conflicting_cols:
+            return (
+                (f"{col[0]}{suffix}", *col[1:])
+                if isinstance(col, tuple)
+                else f"{col}{suffix}"
+            )
+        else:
+            return col
+
+    left_renamer: Dict[IndexLabel, IndexLabel] = {}
+    right_renamer: Dict[IndexLabel, IndexLabel] = {}
+    new_left: List[IndexLabel] = []
+    new_right: List[IndexLabel] = []
+
+    for col in left:
+        new_name = _get_new_name(col, suffixes[0])
+        new_left.append(new_name)
+        left_renamer[col] = new_name
+
+    for col in right:
+        if col not in left_on or col not in right_on:
+            new_name = _get_new_name(col, suffixes[1])
+            new_right.append(new_name)
+            right_renamer[col] = new_name
+
+    new_columns = pandas.Index(new_left).append(pandas.Index(new_right))  # type: ignore # `Index.append` is untyped
```

Comment:
that's the first time I see/use this double-comment syntax, but it seems to be the only way to make human-readable comments for `type: ignore` annotations:
https://github.com/python/mypy/issues/12299#issuecomment-1067454646